target/linux/realtek/files-5.10/Documentation/devicetree/bindings/realtek,otto-timer.yaml

```diff
@@ -0,0 +1,138 @@
+# SPDX-License-Identifier: GPL-2.0
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/timer/realtek,rtl8300-timer.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: Realtek Timer Device Tree Bindings
+
+maintainers:
+  - Markus Stockhausen <markus.stockhausen@gmx.de>
+
+description: |
+  The Realtek SOCs of the RTL83XX and RTL93XX series have 5-7 timers. Their
+  speed is derived from the Lexra Bus (LXB) by dividers. Each timer has a block
+  of 4 control registers in the adress range 0xb800xxxx with following start
+  offsets.
+
+  RTL838X: 0x3100, 0x3110, 0x3120, 0x3130, 0x3140
+  RTL839X: 0x3100, 0x3110, 0x3120, 0x3130, 0x3140
+  RTL930X: 0x3200, 0x3210, 0x3220, 0x3230, 0x3240, 0x3250
+  RTL931X: 0x3200, 0x3210, 0x3220, 0x3230, 0x3240, 0x3250, 0x3270
+
+  The first two timers are connected to an interrupt line. The driver takes the
```

Comment:
It took me about half a day to understand what's going on here. Spoiler: it's Realtek shenanigans again!

When any of the timers are started with interrupts enabled, wrap-around will cause the interrupt pending bit to be set in the timer's INT registers *and* a corresponding bit will be set in the SoC's interrupt controller. This follows the (known) mapping:
* TC0: IRQ 29
* TC1: IRQ 28
* TC2: IRQ 17
* TC3: IRQ 16
* TC4: IRQ 15

This matches the following definitions from the XGS1250-12 SDK:

```c
#define RTL8380_GLBL_INTR_MSK_ADDR                   (0x3000)
  #define RTL8380_GLBL_INTR_MSK_TC0_IE_OFFSET        (29)
  #define RTL8380_GLBL_INTR_MSK_TC1_IE_OFFSET        (28)
  #define RTL8380_GLBL_INTR_MSK_TC2_IE_OFFSET        (17)
  #define RTL8380_GLBL_INTR_MSK_TC3_IE_OFFSET        (16)
  #define RTL8380_GLBL_INTR_MSK_TC4_IE_OFFSET        (15)
```

However, putting one of TIMER2-TIMER4 as the first timer in the list caused hangs on boot, so that obviously didn't work (as you probably found out too).

The SDK also has the following definitions:
```c
#define RTL8380_INTR_ROUTINE_0_ADDR                  (0x3008)
  #define RTL8380_INTR_ROUTINE_0_TC0_RS_OFFSET       (20)
  #define RTL8380_INTR_ROUTINE_0_TC0_RS_MASK         (0xF << RTL8380_INTR_ROUTINE_0_TC0_RS_OFFSET)
  #define RTL8380_INTR_ROUTINE_0_TC1_RS_OFFSET       (16)
  #define RTL8380_INTR_ROUTINE_0_TC1_RS_MASK         (0xF << RTL8380_INTR_ROUTINE_0_TC1_RS_OFFSET)
```
IRR0 has routing indices 31-24. This puts the routing indices for TC0 and TC1 at 29 and 28, matching the earlier `IE_OFFSET` values.

The "fun" begins when one then looks at what the TC2-TC4 values are:
```c
#define RTL8380_INTR_ROUTINE_2_ADDR                  (0x3010)
  #define RTL8380_INTR_ROUTINE_2_TC2_RS_OFFSET       (28)
  #define RTL8380_INTR_ROUTINE_2_TC3_RS_OFFSET       (24)
  #define RTL8380_INTR_ROUTINE_2_TC4_RS_OFFSET       (20)
```
IRR2 has routing indices 15-8 (see where this is going), putting {TC2, TC3, TC4} at routing indices {15, 14, 13}. When putting this translation into the IRQ driver, TC4 wakes up the CPU as expected (TC2 and TC3 untested for now).

TL;DR: The interrupt controller is going to require a table mapping IRR offsets to IMR/ISR offsets. Thanks, Realtek.

This means you are able to specify an interrupt line for every timer. I would suggest to specify the ISR offsets as the canonical HW IRQ numbers, and keep the routing table index mapping internal to the IRQ driver. The latter can be implemented later, as TC0 and TC1 are mapped as expected, and are the only ones needed for now.

Reply:
Writing to the registers at 0x3250-0x325f doesn't have any effect (XGS1250-12, RTL9302B), so it appears TC5 isn't actually there. The watchdog at 0x3260 is correct though, so looks like the left a gap for TC5, but didn't actually include it.

Reply:
Since there's seems to be no IRQ for it; I'm not surprised at all. Could be as simple as miss-communication between the software/spreadsheet guys and the FPGA guys, where the later removed it to save space; but software was 'already working'.